reactos/drivers/usb/usbhub_new/usbhub.c

```diff
@@ -0,0 +1,5074 @@
+        {
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+
+        RtlZeroMemory(Buffer, Length);
+    }
+
+    Urb = ExAllocatePoolWithTag(NonPagedPool,
+                                sizeof(struct _URB_CONTROL_VENDOR_OR_CLASS_REQUEST),
+                                USB_HUB_TAG);
+
+    if (!Urb)
+    {
+        if (Buffer)
+        {
+            ExFreePoolWithTag(Buffer, USB_HUB_TAG);
+        }
+
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    RtlZeroMemory(Urb, sizeof(struct _URB_CONTROL_VENDOR_OR_CLASS_REQUEST));
+
+    if (IsDeviceToHost)
+    {
+        if (BufferLen)
+        {
+            RtlZeroMemory(TransferBuffer, BufferLen);
+        }
+
+        TransferFlags = USBD_TRANSFER_DIRECTION_IN | USBD_SHORT_TRANSFER_OK;
+    }
+    else
+    {
+        if (BufferLen)
+        {
+            RtlCopyMemory(Buffer, TransferBuffer, BufferLen);
+        }
+
+        TransferFlags = USBD_TRANSFER_DIRECTION_OUT;
+    }
+
+    Urb->Hdr.Length = sizeof(struct _URB_CONTROL_VENDOR_OR_CLASS_REQUEST);
+    Urb->Hdr.Function = Function;
+    Urb->Hdr.UsbdDeviceHandle = NULL;
+
+    Urb->TransferFlags = TransferFlags;
+    Urb->TransferBuffer = BufferLen != 0 ? Buffer : NULL;
+    Urb->TransferBufferLength = BufferLen;
+    Urb->TransferBufferMDL = NULL;
+    Urb->UrbLink = NULL;
+
+    Urb->RequestTypeReservedBits = RequestType.B;
+    Urb->Request = Request;
+    Urb->Value = RequestValue;
+    Urb->Index = RequestIndex;
+
+    Status = USBH_FdoSyncSubmitUrb(HubExtension->Common.SelfDevice, (PURB)Urb);
+
+    if (IsDeviceToHost && BufferLen)
+    {
+        RtlCopyMemory(TransferBuffer, Buffer, BufferLen);
+    }
+
+    if (Buffer)
+    {
+        ExFreePoolWithTag(Buffer, USB_HUB_TAG);
+    }
+
+    ExFreePoolWithTag(Urb, USB_HUB_TAG);
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_SyncResetPort(IN PUSBHUB_FDO_EXTENSION HubExtension,
+                   IN USHORT Port)
+{
+    USBHUB_PORT_STATUS PortStatus;
+    KEVENT Event;
+    LARGE_INTEGER Timeout;
+    ULONG ResetRetry = 0;
+    NTSTATUS Status;
+
+    DPRINT("USBH_SyncResetPort: Port - %x\n", Port);
+
+    InterlockedIncrement(&HubExtension->PendingRequestCount);
+
+    KeWaitForSingleObject(&HubExtension->HubPortSemaphore,
+                          Executive,
+                          KernelMode,
+                          FALSE,
+                          NULL);
+
+    Status = USBH_SyncGetPortStatus(HubExtension,
+                                    Port,
+                                    &PortStatus,
+                                    sizeof(USBHUB_PORT_STATUS));
+
+    if (NT_SUCCESS(Status) &&
+        (PortStatus.UsbPortStatus.Usb20PortStatus.CurrentConnectStatus == 0))
+    {
+        Status = STATUS_UNSUCCESSFUL;
+        goto Exit;
+    }
+
+    HubExtension->HubFlags |= USBHUB_FDO_FLAG_RESET_PORT_LOCK;
+    KeInitializeEvent(&Event, NotificationEvent, FALSE);
+
+    while (TRUE)
+    {
+        BM_REQUEST_TYPE RequestType;
+
```

Comment:
usbhub.c:497: please move the KeInitializeEvent here instead of having it both above and below. That's more logical.